assets/js/components/Dashboard/NewPaMessage/StationsAndZones/SelectZonesPage.tsx

```diff
@@ -0,0 +1,541 @@
+      fp.uniq,
+      fp.groupBy((routeID: string) => {
+        if (routeID.startsWith("Green")) {
+          return "Green";
+        }
+
+        if (routeToRouteIDMap["Bus"].includes(routeID)) {
+          return "Bus";
+        }
+
+        if (routeToRouteIDMap["Silver"].includes(routeID)) {
+          return "Silver";
+        }
+
+        return routeID;
+      }),
+    )(places);
+
+    setSelectedRoutes(routes);
+    setSelectedRouteFilter(Object.keys(routes)[0]);
+    setNewPlaces(
+      places.filter((p) =>
+        places
+          .filter((p) => p.screens.some((s) => value.includes(s.id)))
+          .map((p) => p.id)
+          .includes(p.id),
+      ),
+    );
+  }, [places]);
+
+  const filteredPlaces = useMemo(() => {
+    return newPlaces.filter((p) =>
+      p.screens.some((s) =>
+        s.route_ids?.some((r) =>
+          routeToRouteIDMap[selectedRouteFilter]?.some((a) => r === a),
+        ),
+      ),
+    );
+  }, [selectedRouteFilter]);
+
+  const allScreens = useMemo(
+    () =>
+      filteredPlaces.flatMap((p) => {
+        return p.screens.filter(
+          (s) =>
+            fp.intersection(routeToRouteIDMap[selectedRouteFilter], s.route_ids)
+              .length > 0,
+        );
+      }),
+    [filteredPlaces],
+  );
+
+  const screensByZone = useMemo(() => {
+    const screensWithAZone = allScreens.filter((s) => s.zone);
+    return fp.groupBy((s) => s.zone, screensWithAZone);
+  }, [allScreens]);
+
+  const areAllGroupsSelected = (groups: Array<ReadonlyArray<string>>) =>
+    groups.every((group) =>
+      group.every((zone) => {
+        return signIDs(screensByZone[zone] ?? []).every(isSelected);
+      }),
+    );
+
+  const selectGroups = (groups: Array<ReadonlyArray<string>>) => {
+    const signsWithoutSelectedRoute = fp.without(signIDs(allScreens), value);
+
+    const newScreens = groups.flatMap((group) =>
+      group.flatMap((zone) => signIDs(screensByZone[zone] ?? [])),
+    );
+
+    onChange(fp.uniq([...signsWithoutSelectedRoute, ...newScreens]));
+  };
+
+  const isAllSelected =
+    value.length > 0 &&
+    areAllGroupsSelected([LEFT_ZONES, MIDDLE_ZONES, RIGHT_ZONES]);
+  const isLeftSelected =
+    !isAllSelected && areAllGroupsSelected([LEFT_ZONES, MIDDLE_ZONES]);
+  const isRightSelected =
+    !isAllSelected && areAllGroupsSelected([MIDDLE_ZONES, RIGHT_ZONES]);
+
+  const hasRailFilter = () =>
+    [
+      "Green",
+      "Green-B",
+      "Green-C",
+      "Green-D",
+      "Green-E",
+      "Blue",
+      "Red",
+      "Orange",
+      "Mattapan",
+    ].includes(selectedRouteFilter);
+
+  const massSelectClassName = (selected: boolean) =>
+    selected ? "button-primary" : "button-secondary-outline";
+
+  const getFilterPillLabel = () => {
+    if (selectedRouteFilter.startsWith("Green-")) {
+      return `Green Line ${selectedRouteFilter.split("-")[1]}`;
+    } else {
+      return `${selectedRouteFilter} Line`;
+    }
+  };
```

Comment:
**note:** another thunk here...

Reply:
When this was just an if with an else like it is here, I agree that this was unnecessary. I need to add an additional conditional so a simple ternary won't cut it. I could do a nested ternary but that feels harder to read. Do you have a preference for a conditional like this?

Reply:
You could use an [IIFE](https://developer.mozilla.org/en-US/docs/Glossary/IIFE) or `useMemo`.